### PR TITLE
Add support for pluggable explainer runtimes

### PR DESCRIPTION
**Pull request by TimKleinloog**
**What this PR does / why we need it**:
We want to make explainer pluggable as predictors so users can easily add explainer runtimes.
For additional context see https://github.com/kserve/kserve/issues/3398

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Run `make deploy-dev` and deploy the new explainer sample that uses an explainer runtime `docs/samples/v1beta1/explanation/deeploy-shap/deeploy.yaml`

**Special notes for your reviewer**:

1. Some changes include naming decisions: ModelSpec is reused for ExplainerSpec, how to both use PredictorExtensionSpec and ExplainerExtensionSpec in ModelSpec?
2. Some test are failing for auto generated code, how to deal with that?

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
You can now configure a model in the ExplainerSpec to plug any arbitrary framework
```

### CODE REVIEW

File: install/v0.12.0-rc1/kserve-cluster-resources.yaml [hunk 1]
```diff
@@ -338,6 +338,40 @@ spec:
     version: "1"
 ---
 apiVersion: serving.kserve.io/v1alpha1
+kind: ClusterServingRuntime
```
Comment (terrytangyuan) [inline install/v0.12.0-rc1/kserve-cluster-resources.yaml]: This folder is for `v0.12.0-rc1` which is already released. You can remove this

Comment (TimKleinloog) [reply install/v0.12.0-rc1/kserve-cluster-resources.yaml]: What folder can I add changes to installation files?

Comment (terrytangyuan) [reply install/v0.12.0-rc1/kserve-cluster-resources.yaml]: It will be done as part of the release process

**Comment by yuzisun** [comment]
@TimKleinloog can you help rebase from master, there was a fix for the CI failures. Thanks!

**Comment by TimKleinloog** [comment]
> @TimKleinloog can you help rebase from master, there was a fix for the CI failures. Thanks!

Thanks for letting me know, just rebased

**Review by terrytangyuan — COMMENTED** [review]
You'll need to run `make generate` to fix codegen build.

**Comment by TimKleinloog** [comment]
> You'll need to run `make generate` to fix codegen build.

Thanks @terrytangyuan , fyi in my case `make generate` actually made the test fail, running `make manifests` again fixed it.

The remaining failing test are likely caused by a exceeding 262144 characters in last-applied-configuration for the CustomResourceDefinition "inferenceservices.serving.kserve.io". Any advice how to deal with that?

**Comment by terrytangyuan** [comment]
> The remaining failing test are likely caused by a exceeding 262144 characters in last-applied-configuration for the CustomResourceDefinition "inferenceservices.serving.kserve.io". Any advice how to deal with that?

I just saw it in one of my PRs as well so it may not be related to your changes.

**Comment by terrytangyuan** [comment]
Actually it's related. Both your PR and my PR makes `config/crd/serving.kserve.io_inferenceservices.yaml` much larger which exceeds the limit.

**Comment by terrytangyuan** [comment]
Looks like it's only complaining about `metadata.annotations` in my PR. 

> The remaining failing test are likely caused by a exceeding 262144 characters in last-applied-configuration for the CustomResourceDefinition "inferenceservices.serving.kserve.io"

Where did you observe that it's caused by `last-applied-configuration`?

**Comment by TimKleinloog** [comment]
> Looks like it's only complaining about `metadata.annotations` in my PR.
> 
> > The remaining failing test are likely caused by a exceeding 262144 characters in last-applied-configuration for the CustomResourceDefinition "inferenceservices.serving.kserve.io"
> 
> Where did you observe that it's caused by `last-applied-configuration`?

I saw a similar issue [here](https://github.com/argoproj/argo-cd/issues/820), that was related to the `kubectl.kubernetes.io/last-applied-configuration` and yaml size. Do you think we are facing something similar here?

**Comment by terrytangyuan** [comment]
We discussed briefly at the community meeting. The problem is that the CRD is too large and the error message could be misleading. Tracking in https://github.com/kserve/kserve/issues/3487

**Pull request merged**